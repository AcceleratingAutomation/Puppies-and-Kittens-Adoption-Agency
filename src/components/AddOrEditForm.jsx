import React from "react";
import PropTypes from "prop-types";
import {
  Grid,
  Container,
  FormControl,
  FormHelperText,
  Select,
  MenuItem,
} from "@material-ui/core";
import { Form, useFormikContext } from "formik";
import { FormField, MultiLineFormField } from "./FormField";
import FormButtons from "./FormButtons";
import { bioLabel } from "../accessibility/users/usersText";

function AddOrEditForm({ errors, touched, formFields, endpoint, id }) {
  const { values, handleChange } = useFormikContext();

  return (
    <Form>
      <Container>
        <Grid
          container
          justify="center"
          alignItems="flex-start"
          direction="row"
        >
          <Grid item xs={12} sm={6} style={{ textAlign: "left" }}>
            {formFields.map((field) =>
              field.type === "select" ? (
                <Grid container alignItems="center" spacing={2} key={field.id}>
                  <Grid item>
                    <label htmlFor="name" style={{ fontWeight: "bold" }}>
                      {field.label}
                    </label>
                  </Grid>
                  <Grid item>
                    <FormControl
                      key={field.id}
                      error={touched[field.name] && Boolean(errors[field.name])}
                    >
                      <Select
                        id={field.id}
                        name={field.name}
                        value={values[field.name]}
                        onChange={handleChange}
                      >
                        {field.options.map((option) => (
                          <MenuItem key={option} value={option}>
                            {option}
                          </MenuItem>
                        ))}
                      </Select>
                      {touched[field.name] && errors[field.name] && (
                        <FormHelperText>{errors[field.name]}</FormHelperText>
                      )}
                    </FormControl>
                  </Grid>
                </Grid>
              ) : (
                <FormField
                  key={field.id}
                  id={field.id}
                  name={field.name}
                  type={field.type}
                  label={field.label}
                  errors={errors}
                  touched={touched}
                />
              ),
            )}
          </Grid>
          <Grid item xs={12} sm={6} style={{ textAlign: "left" }}>
            <MultiLineFormField
              id="bio"
              name="bio"
              label={bioLabel}
              errors={errors}
              touched={touched}
            />
          </Grid>
        </Grid>
        <FormButtons id={id} endpoint={endpoint} />
      </Container>
    </Form>
  );
}

AddOrEditForm.propTypes = {
  errors: PropTypes.instanceOf(Object).isRequired,
  touched: PropTypes.instanceOf(Object).isRequired,
  formFields: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      options: PropTypes.arrayOf(PropTypes.string),
    }),
  ).isRequired,
  endpoint: PropTypes.string.isRequired,
  id: PropTypes.string,
};

AddOrEditForm.defaultProps = {
  id: "",
};

export default AddOrEditForm;
