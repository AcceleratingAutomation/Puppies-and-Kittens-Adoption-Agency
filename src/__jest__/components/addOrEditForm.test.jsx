import React from "react";
import { render, screen } from "@testing-library/react";
import { Formik } from "formik";
import AddOrEditForm from "../../components/AddOrEditForm";

function renderAddOrEditForm(props) {
  return render(
    <Formik initialValues={{ type: "" }} onSubmit={jest.fn()}>
      <AddOrEditForm
        errors={props.errors}
        touched={props.touched}
        formFields={props.formFields}
        endpoint={props.endpoint}
        id={props.id}
      />
    </Formik>,
  );
}

describe("Add Or Edit Form", () => {
  const formProps = {
    errors: {},
    touched: {},
    formFields: [
      {
        id: "1",
        name: "name",
        type: "text",
        label: "Name",
      },
      {
        id: "2",
        name: "type",
        type: "select",
        label: "Type",
        options: ["Option 1", "Option 2"],
      },
    ],
    endpoint: "/test-endpoint",
    id: "1",
  };

  beforeEach(() => {
    renderAddOrEditForm(formProps);
  });

  test("matches snapshot comparison", () => {
    const { asFragment } = renderAddOrEditForm(formProps);
    expect(asFragment()).toMatchSnapshot();
  });

  test("displays form buttons", () => {
    expect(screen.getByText("Save")).toBeInTheDocument();
    expect(screen.getByText("Cancel")).toBeInTheDocument();
  });

  test("displays error message when field is touched and there is an error", () => {
    const errorProps = {
      ...formProps,
      errors: { type: "Type error" },
      touched: { type: true },
    };
    renderAddOrEditForm(errorProps);
    expect(screen.getByText("Type error")).toBeInTheDocument();
  });
});
