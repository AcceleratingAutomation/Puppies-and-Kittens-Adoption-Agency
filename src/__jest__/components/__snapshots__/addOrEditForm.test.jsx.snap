// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Or Edit Form matches snapshot comparison 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-flex-start MuiGrid-justify-xs-center"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
          style="text-align: left;"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <label
              for="1"
            >
              <strong>
                Name
              </strong>
              <input
                id="1"
                name="name"
                style="margin: 0.625rem;"
                type="text"
                value=""
              />
            </label>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-align-items-xs-center"
          >
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <label
                for="name"
                style="font-weight: bold;"
              >
                Type
              </label>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiFormControl-root"
              >
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                >
                  <div
                    aria-haspopup="listbox"
                    aria-labelledby="2"
                    class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                    id="2"
                    role="button"
                    tabindex="0"
                  >
                    <span>
                      ​
                    </span>
                  </div>
                  <input
                    aria-hidden="true"
                    class="MuiSelect-nativeInput"
                    name="type"
                    tabindex="-1"
                    value=""
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSelect-icon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
          style="text-align: left;"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <label
              for="bio"
            >
              <strong>
                Bio
              </strong>
              <div
                class="MuiFormControl-root MuiTextField-root"
                style="margin: 0.625rem; width: 100%;"
              >
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
                    id="bio"
                    name="bio"
                    rows="10"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
                    style="padding-left: 8px;"
                  >
                    <legend
                      class="PrivateNotchedOutline-legend-2"
                      style="width: 0.01px;"
                    >
                      <span>
                        ​
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </label>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-justify-xs-center"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge"
          style="margin: 0.625rem;"
          tabindex="0"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Save
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge"
          style="margin: 0.625rem;"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Cancel
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
